src/coreclr/jit/rationalize.cpp

```diff
@@ -37,57 +38,143 @@ void Rationalizer::RewriteNodeAsCall(GenTree**             use,
+        {
+            assert(returnType == TYP_UNKNOWN);
+            call->gtCallMoreFlags |= GTF_CALL_M_RETBUFFARG;
+        }
     }
 
-    if (arg1 != nullptr)
+    assert(retType == tree->gtType);
+
+    CORINFO_ARG_LIST_HANDLE sigArg   = sig.args;
+    size_t                  firstArg = 0;
+
+    if (sig.hasThis())
     {
-        call->gtArgs.PushFront(comp, NewCallArg::Primitive(arg1));
-        call->gtFlags |= arg1->gtFlags & GTF_ALL_EFFECT;
+        GenTree*   operand = operands[0];
+        NewCallArg arg     = NewCallArg::Primitive(operand);
+
+        call->gtArgs.PushBack(comp, arg);
+        call->gtFlags |= operand->gtFlags & GTF_ALL_EFFECT;
+
+        firstArg++;
     }
 
-#if DEBUG
-    CORINFO_SIG_INFO sig;
-    comp->eeGetMethodSig(callHnd, &sig);
-    assert(JITtype2varType(sig.retType) == tree->gtType);
-#endif // DEBUG
+    for (size_t i = firstArg; i < operandCount; i++)
+    {
+        GenTree* operand = operands[i];
+
+        CORINFO_CLASS_HANDLE clsHnd = NO_CLASS_HANDLE;
+        CorInfoType          corTyp = strip(comp->info.compCompHnd->getArgType(&sig, sigArg, &clsHnd));
+        var_types            sigTyp = JITtype2varType(corTyp);
+
+        NewCallArg arg;
+
+        if (varTypeIsStruct(sigTyp))
+        {
+            sigTyp = comp->impNormStructType(clsHnd);
+            arg    = NewCallArg::Struct(operand, sigTyp, clsHnd);
+        }
+        else
+        {
+            arg = NewCallArg::Primitive(operand, sigTyp);
+        }
+
+        call->gtArgs.PushBack(comp, arg);
+        call->gtFlags |= operand->gtFlags & GTF_ALL_EFFECT;
 
-#ifdef FEATURE_READYTORUN
+        sigArg = comp->info.compCompHnd->getArgNext(sigArg);
+    }
+
+#if defined(FEATURE_READYTORUN)
     call->AsCall()->setEntryPoint(entryPoint);
-#endif
+#endif // FEATURE_READYTORUN
+
+    unsigned tmpNum = BAD_VAR_NUM;
+
+    if (call->TreatAsShouldHaveRetBufArg())
+    {
+        assert(call->ShouldHaveRetBufArg());
+
+        tmpNum = comp->lvaGrabTemp(true DEBUGARG("return buffer for hwintrinsic"));
+        comp->lvaSetStruct(tmpNum, sig.retTypeClass, false);
+
+        GenTree*   destAddr = comp->gtNewLclVarAddrNode(tmpNum, TYP_BYREF);
+        NewCallArg newArg   = NewCallArg::Primitive(destAddr).WellKnown(WellKnownArg::RetBuffer);
+
+        call->gtArgs.InsertAfterThisOrFirst(comp, newArg);
+        call->gtType = TYP_VOID;
+    }
 
     call = comp->fgMorphArgs(call);
 
+    GenTree* result = call;
+
     // Replace "tree" with "call"
     if (parents.Height() > 1)
     {
-        parents.Top(1)->ReplaceOperand(use, call);
+        if (tmpNum != BAD_VAR_NUM)
+        {
+            result = comp->gtNewLclvNode(tmpNum, tree->gtType);
+        }
+
+        parents.Top(1)->ReplaceOperand(use, result);
+
+        if (tmpNum != BAD_VAR_NUM)
+        {
+            comp->gtSetEvalOrder(result);
+            BlockRange().InsertAfter(insertionPoint, LIR::Range(comp->fgSetTreeSeq(result), result));
```

Comment:
Does `LIR::SeqTree` not work here?

Reply:
I was just following the existing pattern that `RewriteNodeAsCall` had already been using.

As far as I can see, `LIR::SeqTree` is mostly only called from `lower.cpp`, but looks like it'd just be `BlockRange().InsertAfter(insertionPoint, LIR::SeqTree(comp, result))`?

Reply:
Yeah, that looks right to me.

Reply:
Hmmm, it looks like the reason this wasn't being done is because we're doing:
```csharp
compiler->gtSetEvalOrder(tree);
LIR::Range(compiler->fgSetTreeSeq(tree, /* isLIR */ false), tree);
```

While `LIR::SeqTree` is doing:
```
comp->gtSetEvalOrder(tree);
LIR::Range(comp->fgSetTreeSeq(tree, /* isLIR */ true), tree);
```

That is, it's a difference of `isLIR` needing to be `false`. My guess is this is because we're still in rationalization and thus in a weird mixed state of HIR vs LIR where the node won't technically be in LIR form until after it's been processed by rationalization (which will happen after `RewriteNodeAsCall` returns and the visitor processes it as part of `WALK_CONTINUE`)